### PR TITLE
Allow __ua_domain__ to be a list of domain strings

### LINKED ISSUE #239: Multiple domain support for backends
This would allow backends to support multiple domains by providing a comma separated list of domains. E.g. `"a,b"` would cover multimethods that would be covered by both `"a"` and `"b"`.

Alternate: Use a list. `["a", "b"]`.

### PR DESCRIPTION
**Pull request by peterbell10**
Closes #239

This allows a backend to implement multiple domains by making `__ua_domain__` a list of domains:
```python
__ua_domain__ = ["domain_a", "domain_b"]
```

TODO:

- [x] More test coverage (`skip_backend`, `register_backend`, etc.)
- [x] Benchmark function call performance. (`context_helper` in particular is used for functions with default implementations)
- [x] Validate `__ua_domain__` upfront before modifying `uarray`'s state.
- [x] Update documentation

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Can't we just add the backend to each domain's `vector` instead of going through all this hula-hoopery of checking the `domain` each time?

**Comment by peterbell10** [comment]
What do you mean?  I do add it to each domain's vector.

**Comment by hameerabbasi** [comment]
Ah, yes. I see. I guess I'm just surprised by the size of the PR.

**Comment by peterbell10** [comment]
Well `set_global_backend`, `register_backend`, `set_backend`, `skip_backend` and `context_helper` all need to be modified to work with lists of domains.

**Comment by peterbell10** [comment]
>Benchmark function call performance. 

In an unexpected turn of events, it appears to actually be faster than before...
For the simple benchmark below, before I get 1.77 us per call on master but now I get around 1.65 us per call. @hameerabbasi could you confirm? I'm on an old machine at the moment so that might be skewing my results.

```python
import uarray as ua

class Backend: 
    __ua_domain__ = 'ua_test' 

    def __ua_function__(self, func, a, kw): 
        if func.__name__ == 'default': 
            return object() 
        return NotImplemented

@ua.create_multimethod(lambda a, kw, d: ((),{}), domain='ua_test') 
def default(): 
    pass

func = ua.generate_multimethod(lambda: (), lambda a, kw, d: (a, kw), 'ua_test', default=lambda: default())

with ua.set_backend(Backend()): 
    %timeit func()
```

**Comment by hameerabbasi** [comment]
> @hameerabbasi could you confirm?

It seems about the same for me, 948 ns vs 923.

File: uarray/tests/test_uarray.py [hunk 1]
```diff
@@ -379,3 +380,50 @@ def __ua_function__(self, f, a, kw):
...
+    assert_backend_active(be)
+
+    for i in range(len(mms)):
+        ua.clear_backends(mms[i].domain, globals=True)
+
+        with pytest.raises(ua.BackendNotImplementedError):
+            mms[i]()
+
+        for j in range(i + 1, len(mms)):
+            assert mms[j]() == be.ret
```
Comment (hameerabbasi) [inline uarray/tests/test_uarray.py:424]:
```suggestion
            assert mms[j]() is be.ret
```

File: uarray/small_dynamic_array.h [hunk 2]
```diff
@@ -0,0 +1,228 @@
+#pragma once
```
Comment (hameerabbasi) [inline uarray/small_dynamic_array.h:1]: We should probably add a copyright note if this is taken from somewhere.

Comment (peterbell10) [reply uarray/small_dynamic_array.h:1]: I wrote it.

**Pull request merged**